File: rplugin/python3/gdb/breakpoint.py
"""."""

from typing import Dict, List
from gdb.common import Common
from gdb.proxy import Proxy
from gdb.backend.gdb import Gdb


class Breakpoint(Common):
    """Handle breakpoint signs."""

    def __init__(self, common: Common, proxy: Proxy, backend: Gdb):
        """ctor."""
        super().__init__(common)
        self.proxy = proxy
        self.backend = backend
        # Discovered breakpoints so far: {file -> {line -> [id]}}
        self.breaks: Dict[str, Dict[str, List[str]]] = {}
        self.max_sign_id = 0

        # Backend class to query breakpoints
        self.impl = self.backend.Breakpoint(self.proxy)

    def clear_signs(self):
        """Clear all breakpoint signs."""
        for i in range(5000, self.max_sign_id + 1):
            self.vim.call('sign_unplace', 'NvimGdb', {'id': i})
        self.max_sign_id = 0

    def _set_signs(self, buf: int):
        if buf != -1:
            sign_id = 5000 - 1
            # Breakpoints need full path to the buffer (at least in lldb)
            bpath = self.vim.call("expand", f'#{buf}:p')

            def _get_sign_name(count):
                max_count = len(self.config.get('sign_breakpoint'))
                idx = count if count < max_count else max_count - 1
                return f"GdbBreakpoint{idx}"

            for line, ids in self.breaks.get(bpath, {}).items():
                sign_id += 1
                sign_name = _get_sign_name(len(ids))
                self.vim.call('sign_place', sign_id, 'NvimGdb', sign_name, buf,
                              {'lnum': line, 'priority': 10})
            self.max_sign_id = sign_id

    def query(self, buf_num: int, fname: str):
        """Query actual breakpoints for the given file."""
        self.logger.info("Query breakpoints for %s", fname)
        self.breaks[fname] = self.impl.query(fname)
        self.clear_signs()
        self._set_signs(buf_num)

    def reset_signs(self):
        """Reset all known breakpoints and their signs."""
        self.breaks = {}
        self.clear_signs()

    def get_for_file(self, fname: str, line: int):
        """Get breakpoints for the given position in a file."""
        breaks = self.breaks.get(fname, {})
        return breaks.get(f"{line}", {})   # make sure the line is a string

File: lua/nvimgdb/breakpoint.lua
-- Handle breakpoint signs.
-- vim: set et ts=2 sw=2:

local log = require'nvimgdb.log'

---@alias FileBreakpoints table<number, string[]>    # breakpoint collection for a file {line -> [id]}
---@alias QueryBreakpoints function(fname: string, proxy: Proxy): FileBreakpoints  # Function to obtain a breakpoint collection

---@class Breakpoint breakpoint signs handler
---@field private config Config resolved configuration
---@field private proxy Proxy connection to the side channel
---@field private query_impl QueryBreakpoints function to query breakpoints for a given file
---@field private breaks table<string, FileBreakpoints> discovered breakpoints so far: {file -> {line -> [id]}}
---@field private max_sign_id number biggest sign identifier for the breakpoints in use
local Breakpoint = {}
Breakpoint.__index = Breakpoint

---Constructor
---@param config Config resolved configuration
---@param proxy Proxy @connection to the side channel
---@param query_impl QueryBreakpoints @function to query breakpoints
---@return Breakpoint @new instance
function Breakpoint.new(config, proxy, query_impl)
  log.debug({"Breakpoint.new", config = config, proxy = proxy, query_impl = query_impl})
  local self = setmetatable({}, Breakpoint)
  self.config = config
  self.proxy = proxy
  self.query_impl = query_impl
  self.breaks = {}
  self.max_sign_id = 0
  return self
end

---Clear all breakpoint signs in all buffers
function Breakpoint:clear_signs()
  log.debug({"Breakpoint:clear_signs"})
  -- Clear all breakpoint signs.
  for i = 5000, self.max_sign_id do
    vim.fn.sign_unplace('NvimGdb', {id = i})
  end
  self.max_sign_id = 0
end

---Set a breakpoint sign in the given buffer
---@param buf number @buffer number
function Breakpoint:_set_signs(buf)
  log.debug({"Breakpoint:_set_signs", buf = buf})
  if buf ~= -1 then
    local sign_id = 5000 - 1
    -- Breakpoints need full path to the buffer (at least in lldb)
    local bpath = vim.fn.expand('#' .. tostring(buf) .. ':p')

    local function _get_sign_name(idx)
      local max_count = #self.config:get('sign_breakpoint')
      if idx > max_count then
        idx = max_count
      end
      return "GdbBreakpoint" .. tostring(idx)
    end

    local priority = self.config:get('sign_breakpoint_priority')
    local for_file = self.breaks[bpath]
    if for_file ~= nil then
      for line, ids in pairs(for_file) do
        if type(line) == "string" then
          sign_id = sign_id + 1
          local sign_name = _get_sign_name(#ids)
          vim.fn.sign_place(sign_id, 'NvimGdb', sign_name, buf,
            {lnum = line, priority = priority})
        end
      end
      self.max_sign_id = sign_id
    end
  end
end

---Query actual breakpoints for the given file.
---@param buf_num number buffer number
---@param fname string full path to the source code file
function Breakpoint:query(buf_num, fname)
  log.info({"Breakpoint:query(", buf_num = buf_num, fname = fname})
  self.breaks[fname] = self.query_impl(fname, self.proxy)
  self:clear_signs()
  self:_set_signs(buf_num)
end

---Reset all known breakpoints and their signs.
function Breakpoint:reset_signs()
  log.debug({"Breakpoint:reset_signs"})
  self.breaks = {}
  self:clear_signs()
end

---Get breakpoints for the given position in a file.
---@param fname string full path to the source code file
---@param line number|string line number
---@return string[] list of breakpoint identifiers
function Breakpoint:get_for_file(fname, line)
  log.debug({"Breakpoint:get_for_file", fname = fname, line = line})
  local breaks = self.breaks[fname]
  if breaks == nil then
    return {}
  end
  local ids = breaks[tostring(line)]   -- make sure the line is a string
  if ids == nil then
    return {}
  end
  return ids
end

return Breakpoint
